Support sitemaps with more than 50,000 items

jeromeroucou opened the pull request:
**What this PR does / why we need it**:

Actual generation sitemap file doesn't take care the numbers of entries. But sitemap specification indicate _"each Sitemap file that you provide must have no more than 50,000 URLs and must be no larger than 50MB (52,428,800 bytes)"_ (see [sitemap.org](https://www.sitemaps.org/protocol.html#index))

**Which issue(s) this PR closes**:

Closes #8936 

**Special notes for your reviewer**:

Sitemap generation is done by `sitemapgen4j` library, which has the advantage of simplifying the code. I also use the new java 8 date formatter `DateTimeFormatter` which is threadsafe.
I make this PR as a draft because issue have been moved in "SPRINT READY" in [IQSS Dataverse Project](https://github.com/orgs/IQSS/projects/34).

**Suggestions on how to test this**:

A new Unit Test have been added. And if the code is lunch into a running Dataverse, you can also call the API (for ex with `curl -X POST http://localhost:8080/api/admin/sitemap`).

**Is there a release notes update needed for this change?**:

No, I don't thinks.

**Additional documentation**:

Installation configuration need to be updated with the new file name of the sitemap.

Preview at https://dataverse-guide--10321.org.readthedocs.build/en/10321/installation/config.html#creating-a-sitemap-and-submitting-it-to-search-engines

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -1745,6 +1745,8 @@ https://demo.dataverse.org/sitemap.xml is the sitemap URL for the Dataverse Proj
 
 Once the sitemap has been generated and placed in the domain docroot directory, it will become available to the outside callers at <YOUR_SITE_URL>/sitemap/sitemap.xml; it will also be accessible at <YOUR_SITE_URL>/sitemap.xml (via a *pretty-faces* rewrite rule). Some search engines will be able to find it at this default location. Some, **including Google**, need to be **specifically instructed** ...
 
+On Dataverse installation with more than 50000 Dataverse collections or datasets, sitemap file name is ``sitemap_index.html``, not ``sitemap.xml``. Be aware in previous steps to use the correct file name corresponding to your installation.
```
I'm confused. Why not use the library in all cases? And avoid having two code paths and two XML files?

jeromeroucou replied:
My note wasn't precise enough.

In [sitemap specification](https://www.sitemaps.org/protocol.html#index), in the case where is more than 50000 URLs and/or the sitemap file is larger than 50MB, we have to use sitemap index file. The name of this file isn't clearly indicate, but in example there is indicate `http://www.yoursite.com/sitemap_index.xml` or `https://example.com/public/sitemap_index.xml` in [Google developers documentation](https://developers.google.com/search/docs/crawling-indexing/sitemaps/large-sitemaps?hl=en).

The library `sitemapgen4j` generate `sitemap_index.xml` name file by default in the case of sitemap index. You can see it [here](https://github.com/dfabulich/sitemapgen4j/blob/master/src/main/java/com/redfin/sitemapgenerator/SitemapGenerator.java#L220).

So in all cases the library is used, only the name of the sitemap file is change.

I'll have to clarify this in the documentation :+1:

pdurbin commented on pom.xml:
```diff
@@ -532,6 +532,11 @@
           <artifactId>java-json-canonicalization</artifactId>
           <version>1.1</version>
         </dependency>
+        <dependency>
```
I'm just putting this comment here in a random place but I think we should have a release note snippet. It's a nice feature to support more than 50,000 datasets!

pdurbin replied:
There is now a snippet. Thanks!

pdurbin reviewed (commented):
Just a quick review. Thanks for the pull request, @jeromeroucou ! ❤️

scolapasta commented:
Hi @jeromeroucou @pdurbin is this set to be QAed? (or does it still need more review / follwoup for the review). 

Also, can we change this from not being a draft?

poikilotherm commented:
Just a general note here: the dependency added has not been updated since 2019 and uses old APIs. At least it doesn't introduce any other transitive dependencies. There are some forks with small patches.

My suggestion: let's create a fork at github.com/gdcc and have our own package at Maven Central. In case we need patches because of bugs or changes at search engines, we would be much quicker to get this done. It also allows us to modernize the Java code over time.

jeromeroucou commented:
Good catch @poikilotherm ! If there is possible to have a specific fork with patches, it could be nice, and I will be happy to change my PR with this dependency.

pdurbin commented:
@poikilotherm not a bad suggestion. So we'd fork https://github.com/dfabulich/sitemapgen4j under https://github.com/gdcc

@scolapasta and others reading this, what do you think?

pdurbin commented:
I see a 👍 from @scolapasta and no objections [in Slack](https://iqss.slack.com/archives/C010LA04BCG/p1709234431007329) so I forked it to https://github.com/gdcc/sitemapgen4j

Now what? Fix it up and release to Maven Central? Do we want a separate issue to track this? @jeromeroucou are you interested in helping?

jeromeroucou commented:
@pdurbin Yes I am interested. And I also think a separate issue on the forked project can be better to review patchs applications with PR

pdurbin commented:
@jeromeroucou great thanks. @poikilotherm is already working away in this "modernize" branch, if you'd like to try building it: https://github.com/gdcc/sitemapgen4j/tree/modernize

poikilotherm commented:
He doesn't even have to. Just enable Maven Central Snaps and point to the new coordinates: io.gdcc:sitemapgen4j:2.0.0-SNAPSHOT

poikilotherm commented:
https://github.com/gdcc/sitemapgen4j has been releases as v2.0.0 to Maven Central. (Might take another hour until available from all mirrors.)

Have fun!

poikilotherm commented:
I have just released a version 2.1.0 that has again more optimizations. Most important: addressed loads of code smells and removed obsolete, no longer supported Google Sitemap Extensions.

scolapasta commented:
@jeromeroucou Is this all ready for QA now? (if so can you change it from draft - I jsut want to be sure there's not something else we're waiting for here)

jeromeroucou commented:
Hi @scolapasta , the PR is not yet ready as it's waiting for the sitemap4j library namespaces fix, referenced by gdcc/sitemapgen4j#18

scolapasta commented:
> Hi @scolapasta , the PR is not yet ready as it's waiting for the sitemap4j library namespaces fix, referenced by [gdcc/sitemapgen4j#18](https://github.com/gdcc/sitemapgen4j/issues/18)

Ah, OK, that makes sense. We'll keep an eye on for when you switch it out of draft state.

poikilotherm commented:
@jeromeroucou I merged your PR and a new SNAPSHOT version is available. Would you mind testing this in Dataverse again before I cut a release?

jeromeroucou commented:
@poikilotherm Thanks for the update ! On my laptop with the latest SNAPSHOT version, unit tests and sitemap generation with some dataverses and datasets work fine. When the new release will be published, I'll mention it on pom.xml file.

poikilotherm commented:
sitemap4j 2.1.2 is released

jeromeroucou commented:
@poikilotherm thanks for you work on the sitemapgen4j ! The `2.1.2` version is now used on pom.xml

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2068,6 +2068,8 @@ https://demo.dataverse.org/sitemap.xml is the sitemap URL for the Dataverse Proj
 
 Once the sitemap has been generated and placed in the domain docroot directory, it will become available to the outside callers at <YOUR_SITE_URL>/sitemap/sitemap.xml; it will also be accessible at <YOUR_SITE_URL>/sitemap.xml (via a *pretty-faces* rewrite rule). Some search engines will be able to find it at this default location. Some, **including Google**, need to be **specifically instructed** ...
 
+According to `Sitemaps.org protocol <https://www.sitemaps.org/protocol.html#index>`_, a sitemap file must have no more than 50,000 URLs and must be no larger than 50MiB. In this case, you have to create sitemap index file witch references multiples sitemap files called ``sitemap_index.xml``, instead of ``sitemap.xml``. Be aware in previous steps to use the correct file name corresponding to your  ...
```
```suggestion
According to `Sitemaps.org protocol <https://www.sitemaps.org/protocol.html#index>`_, a sitemap file must have no more than 50,000 URLs and must be no larger than 50MiB. In this case, the protocol instructs you to create a sitemap index file called ``sitemap_index.xml`` (instead of ``sitemap.xml``), which references multiples sitemap files. Be aware in previous steps to use the correct file name corresponding to your situation.
```

I'm suggesting "situation" but I'm not sure that's right. We're asking the sysadmin to be conscious of if they are above or below 50,000 URLs? Is this easy to check? Can we just always use the style that allows for over 50,000 URLs so the (poor) sysadmin doesn't have to think about it? 😅

jeromeroucou replied:
Thanks @pdurbin for your review :pray:

Do you think the following addition will help sysadmins ?

_As a URL corresponds to a published collection or dataset, it's possible to determine which file to use based on the sum of these elements._

This reminds me to mention that I didn't add datafile support in the sitemap. I considered that this could be done in another issue. But I'm open to other suggestions :)

landreev replied:
I just left a comment under the release note - a quick similar explanation would help here as well, IMO. Just spelling out that a single `sitemap.xml` is being replaced by the sitemap index **which in turn lists the multiple partial sitemap files, `sitemap_1.xml`, `sitemap_2.xml` ... etc.**
And maybe also a word or two explaining that starting with this version of Dataverse this will be done automatically, depending on how many public datasets and collections are hosted in a given instance. 

Seeing how this has been slightly confusing to a reviewer, adding these extra words may help a few remote admins understand what's going on.

landreev replied:
And no, I wouldn't worry about datafiles for now. It's been our policy to discourage googlebot from crawling the files, and nobody has yet requested to change it.

pdurbin reviewed (commented):
I have some questions about how this works. Please see my comments in the review. Thanks!

landreev commented on doc/release-notes/8983-more-than-50000-entries-in-sitemap.md:
```diff
@@ -0,0 +1,3 @@
+The sitemap file generation can handle more than 50,000 entries if needed with the [sitemapgen4j](https://github.com/gdcc/sitemapgen4j) library, maintained by the Global Dataverse Community Consortium.
+
+In this case, the Dataverse Admin API `api/admin/sitemap` create a sitemap index file, called `sitemap_index.xml`, in place of the `sitemap.xml` file. See the config section of the Installation Guide for details.
```
Could you please add just a couple of words here, along the lines of "... in place of the single `sitemap.xml`, which will be split into individual files `sitemap_1.xml`, `sitemap_2.xml` etc. as needed; these files will be listed in the index files and placed in the same directory". I obviously encourage you to try and explain this better and in a more clear way (hence "along the lines of ..."

jeromeroucou replied:
Hi @landreev.

Thanks you for the review. A update of release notes snippet and configuration page have been done, hoping to have taken your feedback.

landreev reviewed (commented):
I left a couple of comments. Looks great otherwise. We have more than 50K public objects in our production, so I've been using a simple script to split the sitemap into segments. But it will help other instances to have it done automatically.

pdurbin commented:
@jeromeroucou I took a close look at the documentation and I'm suggesting a significant rewrite here:

- https://github.com/Recherche-Data-Gouv/dataverse/pull/6

If you like the changes, please feel free to merge. Otherwise, I'm happy to talk about it! 😄

pdurbin commented:
I looked at the code and poked around with the new testHugeSiteMap test. I'm basically ready to approve this and move it to "ready for QA" but I'd like to see the doc PR I made above get merged first.

jeromeroucou commented:
After discussion with @pdurbin, a new commit have been done to fix the sitemap URL in case of `sitemap_index.xml` file. A example below :
```xml
<?xml version="1.0" encoding="UTF-8"?>
<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <sitemap>
    <loc>https://domain.tld/sitemap/sitemap1.xml</loc>
    <lastmod>2024-04-22</lastmod>
  </sitemap>
  <sitemap>
    <loc>https://domain.tld/sitemap/sitemap2.xml</loc>
    <lastmod>2024-04-22</lastmod>
  </sitemap>
</sitemapindex>
```
In `<loc>` tag, if `/sitemap/` isn't present, the location is invalid. There is no "pretty-faces" rewrite for these files (`sitemap1.xml`, ...), unlike `sitemap.xml` and `sitemap_index.xml` files.

I've also modified the unit tests to validate this modification, and I've also added some comments.

landreev commented:
Merging. 
(I checked out and synced the branch with develop locally; but it seems safe to merge without updating it in back in Recherche-Data-Gouv fork... one way to find out).
